libs/@local/harpc/crates/wire/src/request/mod.rs

```diff
@@ -0,0 +1,327 @@
+pub use bytes::Bytes;
+use error_stack::{Result, ResultExt};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use self::{
+    body::{RequestBody, RequestBodyContext},
+    codec::{DecodeError, EncodeError},
+    header::RequestHeader,
+};
+use crate::codec::{Decode, DecodePure, Encode};
+
+pub mod authorization;
+pub mod begin;
+pub mod body;
+pub mod codec;
+pub mod encoding;
+pub mod flags;
+pub mod frame;
+pub mod header;
+pub mod id;
+pub mod procedure;
+pub mod service;
+
+/// A request message.
+///
+/// A request message is composed of a header and a body and is used to send the actual message over
+/// the wire, agnostic over the transport layer.
+///
+/// This is a binary protocol with minimal overhead with framing support. Messages can be up to
+/// 64KiB in size, larger messages are split into multiple frames.
+///
+/// The transport layer is responsible for ensuring that the message is sent and received in order.
+///
+///
+/// # `Begin` Packet
+///
+/// The layout of a `Begin` packet is as follows:
+///
+/// ```text
+/// 0                   1                   2                   3
+/// 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |  Magic  |P|R. |F|S. |S. |P. |E. |A. |      Authorization      |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// |     |P. |                       Payload                       |
+/// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///
+/// * Magic (5 bytes)
+/// * Protocol Version (1 byte)
+/// * Request Id (2 bytes)
```

Comment:
It would be good if every request ID is unique. For that, 2 bytes are not enough.

Reply:
The idea is that the request id is unique *in flight* per session and only used to associate a request with a response. This would only pose a problem if we had more than 65.536 in-flight requests in a single session. Like TCP, this is a rolling number, so once 65.536 is reached, we go back down to 0.

The difference is that TCP uses 32bit instead of 16bit, but the difference is that TCP increments this per packet sent, while we only increment per request sent. So I thought this would be sufficient.

Reply:
I wonder if it would be useful to keep a global request ID for provenance. This ID could be logged e.g. in Sentry. Having a built-in functionality to track a request might be helpful, but I don't feel strongly.

Reply:
65.536 sounds like a lot of in-flight requests and the bottleneck would probably somewhere else, my instinct is that we chose 4 bytes anyway.
In any case we would need a free-list (or an "in-use-list") to find the next free ID as soon as it wraps but that's an implementation detail.

Reply:
We could designate a global request ID on the server for this (which we could return on response), but this request ID is chosen by the client in the request, so it isn't guaranteed to be global. However, we can hand out and designate global IDs from the server, yes.

but I think both that global request id as well as tracing sentry support might be something that's better implemented on the higher abstraction of the payload instead. What do you think? (So that the response/request gain a field for sentry tracing and the response returns a global request id)

Reply:
Unless we increment a global number (-> impossible) we cannot guarantee uniqueness, even with UUIDs. I agree that a tracing identifier is better suited on a higher abstraction layer. The wire protocol should contain the absolute minimum.

How the client would know the request ID before a response of the server is returned?

Reply:
> In any case we would need a free-list (or an "in-use-list") to find the next free ID as soon as it wraps but that's an implementation detail.

Not necessarily; we could be like TCP and assume that once we overflow, assume we're in unused territory. going with 4 bytes to future-proof is a good idea. Especially w/ 2^32 possibilities that's a pretty safe assumption and removes a lot of implementation complexity.

The request id is chosen by the client, and is more like a ticket than anything else, it's basically a: `hey this request is known as A`, then the server is done. This allows for multiple requests on the same band, while not relying on the fact that a will be done after b, the server then replies with: `this is the response to A`. This is commonly used in all sorts of protocols, even JSONRPC.